test/test_nested_tensor_masking.py

```diff
@@ -39,312 +45,995 @@ def test_gen_nested_tensor(self):
-        self.assertEqual(res_mask, exp_mask)
-        self.assertEqual(res_tensor, exp_tensor)
-
-        # mask_dim = 2
-        res_tensor, res_mask = original_nt.to_tensor_mask(mask_dim=2)
-        exp_mask = torch.tensor([[True], [True]])
-        exp_tensor = torch.tensor([[[[[1, 2], [3, 4]]]],
-                                   [[[[5, 6], [7, 8]]]]])
-        self.assertEqual(res_mask, exp_mask)
-        self.assertEqual(res_tensor, exp_tensor)
-        
-        # mask_dim = 3
-        res_tensor, res_mask = original_nt.to_tensor_mask(mask_dim=3)
-        exp_mask = torch.tensor([[[True]], [[True]]])
-        exp_tensor = torch.tensor([[[[[1, 2], [3, 4]]]],
-                                   [[[[5, 6], [7, 8]]]]])
-        self.assertEqual(res_mask, exp_mask)
-        self.assertEqual(res_tensor, exp_tensor)    
-
-        # mask_dim = 4
-        res_tensor, res_mask = original_nt.to_tensor_mask(mask_dim=4)
-        exp_mask = torch.tensor([[[[True, True]]],[[[True, True]]]])
-        exp_tensor = torch.tensor([[[[[1, 2], [3, 4]]]],
-                                   [[[[5, 6], [7, 8]]]]])
-        self.assertEqual(res_mask, exp_mask)
-        self.assertEqual(res_tensor, exp_tensor) 
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[0], [11]], dtype=torch.long))
+        self.assertEqual(mask, torch.tensor([False,  True]))
+        check_tensor_options(self, tensor, a)
 
-    #
-    # Group of tests to test nested_tensor_from_tensor_mask() 
-    #
-    def test_nested_tensor_from_tensor_mask_dim_none(self):
-        original_nt = nt.nested_tensor([
+    def test_scalar_and_empty_nt_cuda(self):
+        a = nt.nested_tensor([
+            nt.nested_tensor([], dtype=torch.long, device='cuda'),
             nt.nested_tensor([
+                torch.tensor(11, dtype=torch.long, device='cuda')
+            ])
+        ])
+
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[0], [11]], dtype=torch.long, device='cuda'))
+        self.assertEqual(mask, torch.tensor([False,  True], device='cuda'))
+        check_tensor_options(self, tensor, a)
+
+    def test_single_tensor(self):
+        a = nt.nested_tensor([
+                torch.tensor([1])
+            ])
+        tensor, mask = a.to_tensor_mask()
+        self.assertEqual(tensor, torch.tensor([[1]]))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+        
+        tensor, mask = a.to_tensor_mask(mask_dim=0)
+        self.assertEqual(tensor, torch.tensor([[1]]))
+        self.assertEqual(mask, torch.tensor(True))
+        check_tensor_options(self, tensor, a)
+        
+        tensor, mask = a.to_tensor_mask(mask_dim=1)
+        self.assertEqual(tensor, torch.tensor([[1]]))
+        self.assertEqual(mask, torch.tensor([True]))
+        check_tensor_options(self, tensor, a)
+        
+        tensor, mask = a.to_tensor_mask(mask_dim=2)
```

Comment:
Shouldn't this fail? The nested_dim of a is 1.

Reply:
but dim is 2. you should be able to have mask all the way to single element of a leaf tensor.